libcontainer/mount_linux.go

```diff
@@ -121,3 +179,75 @@ func syscallMode(i fs.FileMode) (o uint32) {
 	// No mapping for Go's ModeTemporary (plan9 only).
 	return
 }
+
+// mountFd creates an open_tree(2)-like mount fd from the provided
+// configuration. This function must be called from within the container's
+// mount namespace.
+func mountFd(nsHandles *userns.Handles, m *configs.Mount) (mountSource, error) {
+	if !m.IsBind() {
+		return mountSource{}, errors.New("new mount api: only bind-mounts are supported")
+	}
+	if nsHandles == nil {
+		nsHandles = new(userns.Handles)
+		defer nsHandles.Release()
+	}
+
+	var mountFile *os.File
+	var sourceType mountSourceType
+
+	if m.IsBind() {
+		// We only need to use OPEN_TREE_CLONE in the case where we need to use
+		// mount_setattr(2). We are currently in the container namespace and
+		// there is no risk of an opened directory being used to escape the
+		// container. OPEN_TREE_CLONE is more expensive than open(2) because it
+		// requires doing mounts inside a new anonymous mount namespace.
+		if m.IsIDMapped() {
+			flags := uint(unix.OPEN_TREE_CLONE | unix.O_CLOEXEC)
+			if m.Flags&unix.MS_REC == unix.MS_REC {
+				flags |= unix.AT_RECURSIVE
+			}
+			mountFd, err := unix.OpenTree(unix.AT_FDCWD, m.Source, flags)
+			if err != nil {
+				return mountSource{}, &os.PathError{Op: "open_tree(OPEN_TREE_CLONE)", Path: m.Source, Err: err}
+			}
+			mountFile = os.NewFile(uintptr(mountFd), m.Source)
+			sourceType = mountSourceOpenTree
+		} else {
+			var err error
+			mountFile, err = os.OpenFile(m.Source, unix.O_PATH|unix.O_CLOEXEC, 0)
+			if err != nil {
+				return mountSource{}, err
+			}
+			sourceType = mountSourcePlain
+		}
+	}
+
+	if m.IsIDMapped() {
```

Comment:
How about move the code in this block to above(L214)? Then we don't need the next two condition checkes that should never happen.

Reply:
I wrote it this way because it's clearer how it could be extended to support other kinds of mount configuration, but I guess it's better to write it the "obvious" way for now.